source/common/network/happy_eyeballs_connection_impl.cc

```diff
@@ -0,0 +1,433 @@
+#include "source/common/network/happy_eyeballs_connection_impl.h"
+
+#include <vector>
+
+namespace Envoy {
+namespace Network {
+
+HappyEyeballsConnectionImpl::HappyEyeballsConnectionImpl(
+    Event::Dispatcher& dispatcher, const std::vector<Address::InstanceConstSharedPtr>& address_list,
+    Address::InstanceConstSharedPtr source_address, TransportSocketFactory& socket_factory,
+    TransportSocketOptionsConstSharedPtr transport_socket_options,
+    const ConnectionSocket::OptionsSharedPtr options)
+    : dispatcher_(dispatcher), address_list_(address_list), source_address_(source_address),
+      socket_factory_(socket_factory), transport_socket_options_(transport_socket_options),
+      options_(options),
+      next_attempt_timer_(dispatcher_.createTimer([this]() -> void { tryAnotherConnection(); })) {
+  connections_.push_back(createNextConnection());
+}
+
+HappyEyeballsConnectionImpl::~HappyEyeballsConnectionImpl() = default;
+
+void HappyEyeballsConnectionImpl::connect() {
+  ASSERT(!connect_finished_);
+  connections_[0]->connect();
+  maybeScheduleNextAttempt();
+}
+
+void HappyEyeballsConnectionImpl::addWriteFilter(WriteFilterSharedPtr filter) {
+  ASSERT(connect_finished_);
```

Comment:
if we're going to assert these, I think we want API comments in envoy/network/filter.h and add similar asserts to the base connection class to catch slippage

Reply:
Hm. I'm happy to add comments! envoy/network/filter.h is interesting... Connection extends FilterManager (which declares addWriteFilter and friends). But FilterManager doesn't know anything about Connection. So I'm not quite sure how to document this?

Reply:
This is now moot.